Define VK SDK Release Policy (#890)
microsoft/hlsl-specs: proposals/infra/INF-0007-DXC-Vulkan-SDK-Release-Strategy.md

## Proposed solution

> Currently, DXC releases on NuGet, on Windows through VPacks, and on GitHub. There is

Reviewer: nit:  'Windows through VPacks' will read a little ambiguous to readers unfamiliar with the Windows build infrastructure.

Just for your own knowledge,

  VPack is an internal tool we use for package management. The Windows builds ingest specific VPack versions of DXC binaries and source code. The Windows builds have multiple ingestion points for these VPacks for various reasons. For example, there is a VPack that provides the source code to build our HLK tests as part of the OS builds. We also ingest a VPack to use DXC.exe as a build tool for code that compiles shaders as part of its build process.

For Windows, DXC is shipped via the SDK

```suggestion
Currently, DXC releases on NuGet and GitHub. For Windows, DXC is consumed internally through Microsoft's VPack package management system. There is
```

Author: Thanks for clarifying it @alsepkow

infrastructure in place to make sure those releases are happening. The Vulkan SDK
will not modify those releases, it will just append to them.

A Vulkan SDK release should start with LunarG reaching out to the HLSL team, informing them of the
SPIRV-Headers and SPIRV-Tools commits that shall be used to create a release from. From that point
onwards, the Vulkan SDK release is just a process used to build and validate that a
DXC binary is generating valid and correct SPIRV from those submodules.

Validation is done through testing, which is divided into 2 types:
- **Unit Testing**: The DXC codebase contains a series of lit, TAEF and googletests.
- **Functional Tests**: The offload-test repository inside the LLVM org contains the
functional tests built by the HLSL team to validate clang and DXC generated code.

> In order for a release candidate to be considered valid, all tests should pass.

Reviewer: nit: I'm not a fan of saying 'all tests _should_ pass'. Could we elaborate on acceptable reasons when tests are allowed to fail?

## Proposed solution

> Currently, DXC releases on NuGet and GitHub. For Windows, DXC is consumed internally through Microsoft's VPack package management system. There is

Reviewer 2: please rewrap

infrastructure in place to make sure those releases are happening. The Vulkan SDK
will not modify those releases, it will just append to them.

A Vulkan SDK release should start with LunarG reaching out to the HLSL team, informing them of the
SPIRV-Headers and SPIRV-Tools commits that shall be used to create a release from. From that point
onwards, the Vulkan SDK release is just a process used to build and validate that a
DXC binary is generating valid and correct SPIRV from those submodules.

Validation is done through testing, which is divided into 2 types:
- **Unit Testing**: The DXC codebase contains a series of lit, TAEF and googletests.
- **Functional Tests**: The offload-test repository inside the LLVM org contains the
functional tests built by the HLSL team to validate clang and DXC generated code.

In order for a release candidate to be considered valid, all tests should pass.
In case of test failures, the person responsible for the release, along side other 
members of HLSL team, have the autonomy to classify a release candidate valid,

> after proper;y triagging the fail tests.

Reviewer 2: ```suggestion
after determining that shipping with the failed tests is acceptable.
```

## Proposed solution

> Currently, DXC releases on NuGet and GitHub. For Windows, DXC is consumed internally 

Reviewer: ```suggestion
Currently, DXC releases on NuGet, GitHub, and via the Windows SDK. For Windows, DXC is consumed internally 
```

through Microsoft's VPack package management system. There is infrastructure in place 

> to make sure those releases are happening. The Vulkan SDK will not modify those releases, 

Reviewer: 'There is infrastructure in place to make sure those releases are happening' reads ambiguous to me. It's not clear specifically which releases or what this infrastructure really is. It's also unclear to me what it means to 'append' to them.

How about changing to something more like: 

'Currently, DXC releases through GitHub, NuGet, and via the Windows SDK. Within Windows, DXC is consumed through Microsoft's VPack package management system. All of these release channels already have established release processes and automation. The Vulkan SDK release process is additive and does not modify or replace the existing DXC release channels. Instead, it validates a DXC release candidate and produces an additional distribution for inclusion in the Vulkan SDK.'

Reviewer: note that this includes my above suggestion as well.

it will just append to them.

> A Vulkan SDK release should start with LunarG reaching out to the HLSL team, informing them of the

Reviewer: nit
```suggestion
A Vulkan SDK release starts with LunarG reaching out to the HLSL team, informing them of the
```
Are there realistic scenarios where it wouldn't start with LunarG?

Author: No, releasing shader tooling is just a small part of the Vulkan SDK release; that is all coordinated by lunarG. So when they give us the commits other things had to happen on their side before such commits get to us.

> SPIRV-Headers and SPIRV-Tools commits that shall be used to create a release from. From that point

Reviewer: nit: should we link the SPIRV-Headers and SPIRV-Tools repos?

> onwards, the Vulkan SDK release is just a process used to build and validate that a

Reviewer: I'm left a little unclear on what the 'Vulkan SDK release' is from this statement. I'm assuming they post some Vulkan SDK online that includes a DXC binary, yah?

Maybe this should read 'The DXC Vulkan SDK release process' or something similar? We're not really doing a Vulkan SDK release - we're providing a build of DXC that can be used in the Vuklan SDK release?

DXC binary is generating valid and correct SPIRV from those submodules.

> Validation is done through testing, which is divided into 2 types:

Reviewer: Should we elaborate on the platforms where these tests are expected to run? Do they need to pass on linux for   a release to be considered valid?

- **Unit Testing**: The DXC codebase contains a series of lit, TAEF and googletests.

> - **Functional Tests**: The offload-test repository inside the LLVM org contains the

Reviewer: nit: should we link the offload-test repository?

functional tests built by the HLSL team to validate clang and DXC generated code.

> In order for a release candidate to be considered valid, all tests should pass.

Reviewer: nit: I wonder if its worth explicitly stating what a release candidate is? Is it a commit, a GitHub release package, a OneBranch build artifact from a specific commit? This might not be the right place to clarify that. But it's a question that has come up for me a few times now.

In case of test failures, the person responsible for the release, along side other 
members of HLSL team, have the autonomy to classify a release candidate valid,

> after determining that shipping with the failed tests is acceptable.

Reviewer: I wonder if it would be worth it for us to document the 'expected failures' in some way if we have some?

Author: There are no 'expected failures', that is why I added that the team has the autonomy to decide if a test is acceptable to release with or if it should be fixed before that.

Reviewer: Sorry, I meant 'acceptable failures'. But your response still answers my question.

Once a valid candidate is reached, the commit that generated the candidate should

> be tagged and sent back to LunarG, who will be responsible for actually building, packaging

Reviewer: Do we have a policy for what we'll tag it as?

Author: The policy is within the Khronos Group private repository, which is not open source, so it's not linkable from here.

Reviewer: Who decides what it should be tagged with? The policy should be clear on that. Do we need to ask LunarG what the tag should be?

## Proposed solution

Currently, DXC releases through GitHub, NuGet, and via the Windows SDK. Within Windows, 
DXC is consumed through Microsoft's VPack package management system. All of these release 
channels already have established release processes and automation. The Vulkan SDK release 
process is additive and does not modify or replace the existing DXC release channels. 
Instead, it validates a DXC release candidate and produces an additional distribution for 

> inclusion in the Vulkan SDK.

Reviewer: We say that we produce an additional distribution for inclusion in the Vulkan SDK, but further down we say we tag the commit that and share that with LunarG. Do they rebuild from the commit we share or do we provide a 'release package' to them?

Author: They rebuild from the commit we share

> inclusion in the Vulkan SDK.

Reviewer: ```suggestion
DXC releases on NuGet, GitHub, and via the Windows SDK. For Windows, DXC is consumed internally through Microsoft's VPack package management system. These channels already have established release processes and automation. LunarG owns the Vulkan SDK and its release. The HLSL team's role is to *qualify* the build of DXC from a specific commit by validating against the specified SPIRV submodules and tag it, so LunarG can include it in a Vulkan SDK release. This is additive and does not change the existing DXC release channels.
```

The DXC Vulkan SDK release process starts with LunarG reaching out to the HLSL team, informing them of the
[SPIRV-Headers](https://github.com/KhronosGroup/Spirv-headers) and [SPIRV-Tools](https://github.com/KhronosGroup/Spirv-tools) 
commits that shall be used to create a release from. From that point onwards, the Vulkan SDK release is 
just a process used to build and validate that a DXC binary is generating valid and correct SPIRV 
from those submodules. The commit within the DXC repository used to generate such valid binary

> will be considered as a release candidate.

Reviewer: ```suggestion
DXC Vulkan SDK qualification begins when LunarG notifies the HLSL team of the [SPIRV-Headers](https://github.com/KhronosGroup/SPIRV-Headers) and [SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools) commits to build against. The HLSL team builds DXC against those submodules and validates that it generates valid, correct SPIRV. The DXC commit used to produce a validated binary is a *release candidate*.
```

Validation is done through testing, which is divided into 2 types:
- **Unit Testing**: The DXC codebase contains a series of lit, TAEF and googletests.
- **Functional Tests**: The [offload-test repository](https://github.com/llvm/offload-test-suite/)
  inside the LLVM org contains the functional tests built by the HLSL team to validate clang
  and DXC generated code.

In order for a release candidate to be considered valid, all tests should pass.
In case of test failures, the person responsible for the release, along side other 
members of HLSL team, have the autonomy to classify a release candidate valid,
after determining that shipping with the failed tests is acceptable.

Once a valid candidate is reached, the commit that generated the candidate should
be tagged and sent back to LunarG, who will be responsible for actually building, packaging

> and shipping the binary that goes into the Vulkan SDK.

Reviewer: ```suggestion
and shipping the DXC binary that goes into the Vulkan SDK.
```

### Submodule Management

The HLSL team will implement automation to make sure the SPIRV-Headers and SPIRV-Tools
submodules stay up to date. This will allow issues related to such submodules to be detected

> earlier than an actual release date. The details of such automation are not within

Reviewer: ```suggestion
submodules stay up to date. This will allow issues related to such submodules to be detected
before an official qualification run. The details of such automation are not within
```

> the scope of this release.

Reviewer: ```suggestion
the scope of this release strategy.
```

## Proposed solution

> This proposal describes how the SPIRV dependencies are kept current, which dependency

Reviewer 2: > how the SPIRV dependencies are kept current

It looks like the proposal here is to only update when a new Vulkan SDK is going to be built and we're notified of what version to update the submodules to?  This doesn't seem enough, we do need to be continually integrating this, and only lock to the specific versions when we do a release.

Author: The details on how we implement the continuous integration exist in the current version of the spec. I thought the idea was to focus on policies and detail how that current integrates with other DXC releases. Detailing how we detect bugs due to changes on SPIRV seems to be a detail that is not important for the scope of a release and this proposal, IMHO

versions a release is built against, when a build is considered ready, and how the Vulkan
SDK release relates to the other DXC releases. 

The guiding principle is that the Vulkan SDK is not a new release process. It is an
additional consumer of the existing DXC release, with one extra constraint: which
SPIRV-Headers and SPIRV-Tools commits the release is built against. The sections below
first summarize how DXC releases work today, then describe where the Vulkan SDK fits.

### How DXC releases work today

DXC currently ships as a GitHub release, a NuGet package, and Windows/DirectX Vpacks. All
of these are produced from an internal Azure DevOps (ADO) release branch.

A release starts by preparing that branch: the DXC version is bumped, the release notes are
labelled, the branch is cut from main, the shader model is set, and the branch policies are

> locked down. The build, report, and release pipelines are then pointed at the new branch so

Reviewer 2: > build, report, and release pipelines

What are these three pipelines?  I think we have combined build/test/release pipelines for github & vpacks.  Is that detail necessary here anyway?

Author: The idea is to describe the process of DXC release we need to integrate into; the details of it being 1 or 3 pipelines are irrelevant to the scope of this change. What is important is that such a process exists, what it does, and how we can fit within it. I do thing it is important to state the details of DXC build; otherwise, things might be left unclear into where the vulkan release will fit

that subsequent builds run against it.

Submodules are not otherwise part of this process. A submodule stays at whatever commit it
is already set to; advancing SPIRV-Headers and SPIRV-Tools is a manual, optional step, not
something the release does on its own. We assume here that the submodules are left as they

> are unless they are deliberately updated.

Reviewer 2: This seems to be an excessively verbose way of saying something like "Currently, the SPIRV-Headers and SPIRV-Tools submodules are not updated as part of the release process."

Author: Sure, but the paragraph clarifies that no pipeline is intended to handle this; such is the responsibility of whoever will handle the release.

The release builds run from those branches and are marked for retention. The build and
report pipelines that run here are what we today treat as the validation that gates the
release. Once the builds pass, the artifacts are published: the GitHub release is drafted
and published, the NuGet package is uploaded, the Windows and DirectX Vpacks are updated,
the VCPKG mapping is updated, and the new compiler is registered in Compiler Explorer

> (Godbolt).

Reviewer 2: Is this paragraph particularly relevant to this proposal?

Author: They are part of the description of existing DXC release, I think it is

### Where the Vulkan SDK fits

DXC does not build or ship the Vulkan SDK binary. Our deliverable is a validated commit,
not an artifact: once a build passes validation, the commit is tagged as a release
candidate (RC) and that tag is reported to LunarG. LunarG build DXC from that tag and ship
the resulting binary as part of the Vulkan SDK.

The build we validate is the same one that validates every PR — there is no special Vulkan
SDK build. The only addition is a handful of optional stages, created to accommodate the
release — mainly running the offload-tests against the build (see [Validation](#validation)).

The other difference is which SPIRV-Headers and SPIRV-Tools commits the build is built against:
for the Vulkan SDK, the submodules are advanced to the specific commits LunarG specifies
for that SDK before the build is validated. This is a deliberate, manual submodule update. 

### Validation

A commit is tagged as a Vulkan SDK release candidate and reported to LunarG only after it
passes validation against the LunarG-specified SPIRV commits. That validation is the full 
DXC test suite — everything we already run today for a release — plus the offload tests 

> run against lavapipe. 

Reviewer 2: Does this boil down to:

* we've added Vulkan tests via lavapipe + WARP to the DXC CI pipelines
* we'll manually update the SPIRV-Headers and SPIRV-Tools to the appropriate versions for an SDK release
* we'll tag the commit that should included in the Vulkan SDK

Author: Sure

## Proposed solution

This proposal describes how the SPIRV dependencies are kept current, which dependency
versions a release is built against, when a build is considered ready, and how the Vulkan
SDK release relates to the other DXC releases. 

The guiding principle is that the Vulkan SDK is not a new release process. It is an
additional consumer of the existing DXC release, with one extra constraint on which SPIRV
commits the build is pinned to. The sections below first summarize how DXC releases work
today, then describe where the Vulkan SDK fits.

### How DXC releases work today

DXC currently ships as a GitHub release, a NuGet package, and Windows/DirectX Vpacks. The
GitHub release is produced from a release branch in GitHub, while the NuGet package and the

> Vpacks are produced from a release branch in Azure DevOps.

Reviewer 2: All of the releases come from the internal ADO branch.

> Vpacks are produced from a release branch in Azure DevOps.

Reviewer 2: Suggest we refer to this as "internal ADO release branch" or similar.

A release starts by preparing these release branches: the DXC version is bumped, the
release notes are labelled, the branch is cut from main, the shader model is set, and the

> branch policies are locked down. The submodules are pinned on the branch at this point. The

Reviewer 2: > The submodules are pinned on the branch at this point.

Currently we don't think about the submodules at all as part of the release process.  There's no automatic process to update the submodules so it continues to be done manually for main as well as release branches.

> build, submodule-update, report, and release pipelines are then pointed at the new branch

Reviewer 2: > submodule-update

I think this probably refers to updating the submodules on the internal branch, which is what picks up any changes from github.

so that subsequent builds run against it.

The release builds run from those branches and are marked for retention. The build and
report pipelines that run here are what we today treat as the validation that gates the
release. Once the builds pass, the artifacts are published: the GitHub release is drafted
and published, the NuGet package is uploaded, the Windows and DirectX Vpacks are updated,
the VCPKG mapping is updated, and the new compiler is registered in Compiler Explorer
(Godbolt).

### Where the Vulkan SDK fits

> The Vulkan SDK build is the same DXC release described above, with a single added

Reviewer 2: > The Vulkan SDK build is the same DXC release described above

I don't think we're planning on building any binaries for the Vulkan SDK are we?

constraint: it is built against the specific SPIRV-Headers and SPIRV-Tools commits that
LunarG specifies for that SDK, rather than against whatever the branch happened to pin.
We keep the two pinning cases separate:

* Continuous integration builds against the latest SPIRV-Headers and SPIRV-Tools, so
  regressions against upstream are found early. This is independent of any release.

> * A release is built against pinned SPIRV commits. For the Vulkan SDK those commits are

Reviewer 2: I think we should shy away from talking about "pinning" submodules.  To me this suggests that something happens if we don't "pin" them, but that's not the case.  The default behavior for submodules is to stay where they are - we have to go out of our way to update them.
